src/unicycle_2d/ignition.cpp

```diff
@@ -0,0 +1,297 @@
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_rl/unicycle_2d/ignition.h>
+
+#include <fuse_constraints/absolute_constraint.h>
+#include <fuse_core/async_sensor_model.h>
+#include <fuse_core/eigen.h>
+#include <fuse_core/sensor_model.h>
+#include <fuse_core/transaction.h>
+#include <fuse_core/util.h>
+#include <fuse_core/uuid.h>
+#include <fuse_rl/SetPose.h>
+#include <fuse_rl/SetPoseDeprecated.h>
+#include <fuse_variables/acceleration_linear_2d_stamped.h>
+#include <fuse_variables/orientation_2d_stamped.h>
+#include <fuse_variables/position_2d_stamped.h>
+#include <fuse_variables/velocity_angular_2d_stamped.h>
+#include <fuse_variables/velocity_linear_2d_stamped.h>
+#include <fuse_variables/stamped.h>
+
+#include <geometry_msgs/PoseWithCovarianceStamped.h>
+#include <pluginlib/class_list_macros.h>
+#include <std_srvs/Empty.h>
+#include <tf2/convert.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+#include <Eigen/Dense>
+#include <Eigen/Eigenvalues>
+
+#include <exception>
+#include <stdexcept>
+
+
+// Register this motion model with ROS as a plugin.
+PLUGINLIB_EXPORT_CLASS(fuse_rl::unicycle_2d::Ignition, fuse_core::SensorModel);
+
+namespace fuse_rl
+{
+
+namespace unicycle_2d
+{
+
+Ignition::Ignition() :
+  fuse_core::AsyncSensorModel(1),
+  device_id_(fuse_core::uuid::NIL)
+{
+}
+
+void Ignition::onInit()
+{
+  // Read settings from the parameter sever
+  device_id_ = fuse_variables::loadDeviceId(private_node_handle_);
+
+  params_.loadFromROS(private_node_handle_);
+
+  // Connect to the reset service
+  if (!params_.reset_service.empty())
+  {
+    reset_client_ = node_handle_.serviceClient<std_srvs::Empty>(params_.reset_service);
```

Comment:
This requires the config to specify the node name. Unfortunately, `private_node_handle_` also requires that.
I wonder if there's a better solution for this. :thinking:

Reply:
I feel like I need to use the private node handle to advertise topics/services within the plugins. Otherwise the default names will step on each other if you attempt to run the same plugin multiple times (e.g. multiple scan match sensors for robots with multiple lasers). This, of course, puts the node name in the topic and makes remapping topics almost mandatory. Tom's preference was to supply the topic names via parameter instead. Not that it solves anything; the full node name is still required. But it allows you to manage the entire config from within the yaml file, instead of half in the yaml file and half in the launch. I suppose with this method you can choose between using the default topic names and remapping them via the launch file, or providing the desired topic name via parameters.

For this specific line, the optimizer _could_ advertise the "reset" service in the global namespace. It just seemed like "reset" was so generic that it should be advertised in the local namespace. "/smoother/reset" instead of "/reset". But since this service is advertised at the optimizer level and can be easily remapped to whatever you want, I don't have a strong preference on what the default should be.

@ayrton04 Do you have any thoughts on any of this?

Reply:
So yes, the reason I prefer parameters is precisely because plugins are a pain. If you have two plugins, and both subscribe to some topic, and you want to just change _one_ of them to listen to another topic, you're hosed, because remapping will force them both to the new topic. For that reason, I prefer exposing parameters to specify topics. You get the best of both worlds: you can use the parameter, but you can still remap if you want to, as Stephen pointed out.

I think the best option is to advertise locally, subscribe globally. We have to assume that there will be many plugins, so advertising locally make sense. But for subscibers/clients, you can't know the other namespaces in the system, so you have to assume global and remap/use parameters.

Reply:
Thanks for your replies :+1: